src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs

```diff
@@ -131,11 +131,25 @@ protected override void CheckInterfaces(DiagnosticBag diagnostics)
                 var corLibrary = this.ContainingAssembly.CorLibrary;
                 var conversions = new TypeConversions(corLibrary);
                 var location = singleDeclaration.NameLocation;
+                var seenInterfaces = PooledHashSet<NamedTypeSymbol>.GetInstance();
 
                 foreach (var @interface in interfaces)
                 {
                     @interface.CheckAllConstraints(conversions, location, diagnostics);
+
+                    foreach (var seenInterface in seenInterfaces)
+                    {
+                        if (@interface.Equals(seenInterface, TypeCompareKind.IgnoreTupleNames))
```

Comment:
If we're searching the list linearly, we should use an `ArrayBuilder<>` rather than `HashSet<>`.

I think I'd prefer the search to be O(1) though. Consider breaking out the check from the previous loop and only allocating if more than one interface:
```
if (interfaces.Length > 1)
{
    var seenInterfaces = new Dictionary<NamedTypeSymbol, NamedTypeSymbol>(EqualsIgnoringComparer.Instance);
    foreach (var @interface in interfaces)
    {
        NamedTypeSymbol other;
        if (seenInterfaces.TryGetValue(@interface, out other)) { ... }
    }
}
```